File: backend/src/main/java/com/finance/controllers/ReportController.java
package com.finance.controllers;

import com.finance.dto.ReportDtos.AccountBalanceTrendItem;
import com.finance.dto.ReportDtos.CategorySpendResponse;
import com.finance.dto.ReportDtos.FutureBalancePredictionResponse;
import com.finance.dto.ReportDtos.IncomeExpenseTrendItem;
import com.finance.dto.ReportDtos.InsightItem;
import com.finance.services.ReportService;
import java.time.LocalDate;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/reports")
@RequiredArgsConstructor
public class ReportController {

    private final ReportService reportService;

    @GetMapping("/category-spend")
    public CategorySpendResponse categorySpend(
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate startDate,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate endDate
    ) {
        return reportService.categorySpend(startDate, endDate);
    }

    @GetMapping("/income-vs-expense")
    public List<IncomeExpenseTrendItem> incomeVsExpense(
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate startDate,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate endDate
    ) {
        return reportService.incomeExpenseTrend(startDate, endDate);
    }

    @GetMapping("/account-balance-trend")
    public List<AccountBalanceTrendItem> accountBalanceTrend(
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate startDate,
            @RequestParam(required = false) @DateTimeFormat(iso = DateTimeFormat.ISO.DATE) LocalDate endDate
    ) {
        return reportService.accountBalanceTrend(startDate, endDate);
    }

    @GetMapping("/insights")
    public List<InsightItem> insights() {
        return reportService.insights();
    }

    @GetMapping("/future-balance-prediction")
    public FutureBalancePredictionResponse futureBalancePrediction() {
        return reportService.futureBalancePrediction();
    }
}

File: backend-dotnet/FinanceTracker.Api/Controllers/ReportController.cs
using FinanceTracker.Api.Contracts;
using FinanceTracker.Api.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace FinanceTracker.Api.Controllers;

[ApiController]
[Authorize]
[Route("api/reports")]
public sealed class ReportController(ReportService reportService) : ControllerBase
{
    [HttpGet("category-spend")]
    public Task<CategorySpendResponse> CategorySpend([FromQuery] DateOnly? startDate, [FromQuery] DateOnly? endDate, CancellationToken cancellationToken)
    {
        return reportService.CategorySpendAsync(startDate, endDate, cancellationToken);
    }

    [HttpGet("income-vs-expense")]
    public Task<IReadOnlyList<IncomeExpenseTrendItem>> IncomeVsExpense([FromQuery] DateOnly? startDate, [FromQuery] DateOnly? endDate, CancellationToken cancellationToken)
    {
        return reportService.IncomeExpenseTrendAsync(startDate, endDate, cancellationToken);
    }

    [HttpGet("account-balance-trend")]
    public Task<IReadOnlyList<AccountBalanceTrendItem>> AccountBalanceTrend([FromQuery] DateOnly? startDate, [FromQuery] DateOnly? endDate, CancellationToken cancellationToken)
    {
        return reportService.AccountBalanceTrendAsync(startDate, endDate, cancellationToken);
    }

    [HttpGet("insights")]
    public Task<IReadOnlyList<InsightItem>> Insights(CancellationToken cancellationToken)
    {
        return reportService.InsightsAsync(cancellationToken);
    }

    [HttpGet("future-balance-prediction")]
    public Task<FutureBalancePredictionResponse> FutureBalancePrediction(CancellationToken cancellationToken)
    {
        return reportService.FutureBalancePredictionAsync(cancellationToken);
    }

    [HttpGet("trends")]
    public Task<TrendsResponse> Trends([FromQuery] DateOnly? startDate, [FromQuery] DateOnly? endDate, CancellationToken cancellationToken)
    {
        return reportService.TrendsAsync(startDate, endDate, cancellationToken);
    }

    [HttpGet("net-worth")]
    public Task<NetWorthResponse> NetWorth([FromQuery] DateOnly? startDate, [FromQuery] DateOnly? endDate, CancellationToken cancellationToken)
    {
        return reportService.NetWorthAsync(startDate, endDate, cancellationToken);
    }
}
